File: frontend/src/app/services/osrm.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { map } from 'rxjs/operators';

export interface RoutePoint {
  lat: number;
  lng: number;
}

export interface RouteInfo {
  points: RoutePoint[];
  distance: number;  // km
  duration: number;  // minutos
}

@Injectable({
  providedIn: 'root'
})
export class OsrmService {
  private readonly OSRM_URL = 'https://router.project-osrm.org';

  constructor(private http: HttpClient) {}

  getRoute(startLat: number, startLng: number, endLat: number, endLng: number): Observable<RouteInfo> {
    const url = `${this.OSRM_URL}/route/v1/driving/${startLng},${startLat};${endLng},${endLat}?overview=full&geometries=geojson`;
    
    console.log('🌍 [OSRM] Solicitando ruta:', url);
    console.log('🌍 [OSRM] Desde:', startLat, startLng, 'Hasta:', endLat, endLng);
    
    return this.http.get(url).pipe(
      map((response: any) => {
        console.log('📡 [OSRM] Respuesta código:', response.code);
        
        if (response.code === 'Ok' && response.routes && response.routes.length > 0) {
          const route = response.routes[0];
          const geometry = route.geometry.coordinates;
          
          const points: RoutePoint[] = geometry.map((coord: number[]) => ({
            lat: coord[1],
            lng: coord[0]
          }));
          
          console.log(`✅ [OSRM] Ruta obtenida: ${points.length} puntos, ${(route.distance / 1000).toFixed(2)} km, ${(route.duration / 60).toFixed(0)} min`);
          
          return {
            points: points,
            distance: route.distance / 1000,
            duration: route.duration / 60
          };
        }
        
        console.warn('⚠️ [OSRM] No se obtuvo ruta, usando línea recta');
        return {
          points: [{ lat: startLat, lng: startLng }, { lat: endLat, lng: endLng }],
          distance: this.calcularDistancia(startLat, startLng, endLat, endLng),
          duration: this.calcularDistancia(startLat, startLng, endLat, endLng) * 2
        };
      })
    );
  }

  private calcularDistancia(lat1: number, lng1: number, lat2: number, lng2: number): number {
    const R = 6371;
    const dLat = this.deg2rad(lat2 - lat1);
    const dLng = this.deg2rad(lng2 - lng1);
    const a = Math.sin(dLat / 2) * Math.sin(dLat / 2) +
              Math.cos(this.deg2rad(lat1)) * Math.cos(this.deg2rad(lat2)) *
              Math.sin(dLng / 2) * Math.sin(dLng / 2);
    const c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
    return R * c;
  }

  private deg2rad(deg: number): number {
    return deg * (Math.PI / 180);
  }
}

File: mobile/lib/services/osrm_service.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import 'package:latlong2/latlong.dart';
import 'dart:math' as math;

class OSRMService {
  static const String baseUrl = 'https://router.project-osrm.org';

  /// Obtiene la ruta entre dos puntos
  /// Retorna una lista de coordenadas [lat, lng] y la distancia
  static Future<Map<String, dynamic>> getRoute(
    double startLat,
    double startLng,
    double endLat,
    double endLng,
  ) async {
    final url = Uri.parse(
      '$baseUrl/route/v1/driving/$startLng,$startLat;$endLng,$endLat?overview=full&geometries=geojson',
    );

    try {
      final response = await http.get(url);

      if (response.statusCode == 200) {
        final data = jsonDecode(response.body);

        if (data['code'] == 'Ok' && data['routes'].isNotEmpty) {
          final route = data['routes'][0];
          final geometry = route['geometry']['coordinates'];

          // Convertir GeoJSON [lng, lat] a [lat, lng] para flutter_map
          final List<LatLng> points = [];
          for (var coord in geometry) {
            points.add(LatLng(coord[1], coord[0]));
          }

          final distance = route['distance'] / 1000; // Convertir a km
          final duration = route['duration'] / 60; // Convertir a minutos

          return {'points': points, 'distance': distance, 'duration': duration};
        }
      }

      // Fallback: línea recta si OSRM falla
      return _getStraightLine(startLat, startLng, endLat, endLng);
    } catch (e) {
      print('❌ Error OSRM: $e');
      return _getStraightLine(startLat, startLng, endLat, endLng);
    }
  }

  static Map<String, dynamic> _getStraightLine(
    double startLat,
    double startLng,
    double endLat,
    double endLng,
  ) {
    final distance = _calcularDistancia(startLat, startLng, endLat, endLng);
    return {
      'points': [LatLng(startLat, startLng), LatLng(endLat, endLng)],
      'distance': distance,
      'duration': distance * 2, // 2 minutos por km
    };
  }

  static double _calcularDistancia(
    double lat1,
    double lng1,
    double lat2,
    double lng2,
  ) {
    const double R = 6371;
    final dLat = _toRadians(lat2 - lat1);
    final dLng = _toRadians(lng2 - lng1);
    final a =
        math.sin(dLat / 2) * math.sin(dLat / 2) +
        math.cos(_toRadians(lat1)) *
            math.cos(_toRadians(lat2)) *
            math.sin(dLng / 2) *
            math.sin(dLng / 2);
    final c = 2 * math.atan2(math.sqrt(a), math.sqrt(1 - a));
    return R * c;
  }

  static double _toRadians(double degrees) => degrees * math.pi / 180;
}
